### ISSUE TITLE
`TypeVar` constrained to `Literal[True]` and `Literal[False]` is resolved to `bool`

### ISSUE THREAD

**Issue by Viicos** [microsoft/pyright#6917]
Following https://github.com/microsoft/pyright/discussions/6914:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.13&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFADJIzEjkA2NA4sSk0gMYBQ-AlAC8hEhRAAKAEQEZNeo2YsA2gRABXYgF1FDJq1UAxVgGddASkH9eLcmbNQAglM7cQfdTssAuflCBUAAmxMBQAPoRqAxRUhYswDQomiwsvoSiUKYsFpZQALQAfFAAcmDc-kHVUAgOZjbkWa4paSI5eYIgxABuxKwR8AjEUuT5UADEUADKABZgqaFQAEbELqpKhmodujr8QA)

```python
from typing import TypeVar, Literal, Generic

T = TypeVar("T", Literal[True], Literal[False])


class A(Generic[T]):
    def __init__(self, null: T = False) -> None:
        pass


a = A(null=False)

reveal_type(a)  # Should be A[Literal[False]]

```

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by Viicos** [comment]
Thanks for the fix! I see in the PR that you are special casing literals. Should this be added to the spec in some way? (for the record mypy is not inferring the case without an explicit `null` value provided correctly: [playgroud](https://mypy-play.net/?mypy=latest&python=3.12&gist=947b9311a5826e9a6f557673bf86adfc))

**Comment by erictraut** [comment]
> Should this be added to the spec in some way?

Maybe, although in this case I don't think there's any ambiguity because the constraints imposed on the TypeVar (in this case, value constraints) require that the solved type is a literal. The issue with mypy in this case is not because it doesn't solve `T` as a literal but because it doesn't support specifying a default argument for a generic parameter. I'm somewhat surprised that the maintainers of mypy haven't addressed this already since it's one of [the most upvoted mypy feature requests](https://github.com/python/mypy/issues/3737).

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.345, which I just published. It will also be included in a future release of pylance.
